README.md
# Pindah

A tool for writing Android applications in [Mirah](http://mirah.org).

<img src="https://github.com/mirah/pindah/raw/master/pindah.png" alt="Pindah logo" align="right" />

## Requirements
<a name="requirements"></a>

You must have the [Android SDK](http://d.android.com/sdk/) installed with the `tools/` directory placed on your $PATH.

Inside your `tools/` directory, the `android` program will allow you to download "platform" packages corresponding with the Android versions you wish to develop against. You'll need to install "Android SDK Platform Tools" and at least one platform revision to get started. At the time of this writing, Android 2.3.3 (revision 10) and above [covers 100% of the market](http://developer.android.com/resources/dashboard/platform-versions.html) and is a reasonable target for new applications.

Once the platform tools are installed, place the SDK's `platform-tools/` directory on your $PATH.

You will also need the following tools installed before you can develop with Pindah:

* [OpenJDK](http://openjdk.java.net) 6 or 7
* [JRuby](http://jruby.org) 1.7.3 (or above)
* [Mirah](http://github.com/mirah/mirah) 0.1.0 (or above)
* [rake](http://rake.rubyforge.org) 10.0.3 (or above)

### JRuby/Mirah compatability

Pindah has been successfully tested with the following JRuby and Mirah combinations:

| JRuby/Mirah | 0.1.0 | 0.1.1 | 0.1.2 |
| -----------:|:-----:|:-----:|:-----:|
|       1.7.3 | ✓     | x     | x     |
|       1.7.4 | ✓     | ✓     | ✓     |
|       1.7.5 | ✓     | ✓     | ✓     |
|       1.7.6 | ✓     | ✓     | ✓     |
|       1.7.7 | ✓     | ✓     | ✓     |
|       1.7.8 | ✓     | ✓     | ✓     |
|       1.7.9 | ✓     | ✓     | ✓     |
|      1.7.10 | ✓     | ✓     | ✓     |

### Ruby version management

While Linux users can obtain OpenJDK from their distribution, it is recommended that all users utilize something like [rbenv](https://github.com/sstepehenson/rbenv), [rvm](http://rvm.io), or [uru](https://bitbucket.org/jonforums/uru) to manage their Ruby installations and gems:

    # rbenv

    $ rbenv install jruby-1.7.4
    $ rbenv --global jruby-1.7.4
    $ gem install rake mirah

    # rvm

    $ rvm install jruby-1.7.4
    $ gem install rake mirah

## Installation

To install the latest stable version of Pindah, simply install the gem from the command line:

`gem install pindah`

If you want to play with the latest development version, clone the repo onto your local machine:

    $ git clone git://github.com/mirah/pindah
    $ cd pindah/
    $ gem build pindah.gemspec
    $ gem install pindah-0.1.4.dev.gem

## Usage

Similar to other Ruby projects like Rails or Ruboto, the `pindah` executable is responsible for creating and configuring your new Android project(s).  To get started with a simple project, call the `pindah` command like so:

`pindah create org.example.hello_world`

This will create an empty skeleton project with the package name `org.example.hello_world` in the current directory.

You can also configure your project with the following options:

* To create your project in a specific location, specify an optional path for your new project:
  `pindah create org.example.hello_world path_to_project`

* To create and configure a basic initial activity for your project, specify an activity name after the path:
  `pindah create org.example.hello_world path_to_project StartActivity`

All other project management tasks are handled by rake.  For a summary of all possible tasks you can run on your new project, run `rake -T` from your project directory:

    $ rake -T

    rake clean             # Removes output files created by other targets.
    rake compile           # Run compile
    rake debug             # Builds the application and signs it with a debug key.
    rake install           # Installs the newly build package.
    rake installd          # Installs (only) the debug package.
    rake installi          # Installs (only) the instrumented package.
    rake installr          # Installs (only) the release package.
    rake installt          # Installs (only) the test and tested packages.
    rake instrument        # Builds an instrumented packaged.
    rake javac             # Compiles R.java and other gen/ files.
    rake logcat            # Tail logs from a device or a device or emulator
    rake release           # Builds the application in release mode.
    rake release_unsigned  # Builds the application in release mode.
    rake spec              # Print the project spec
    rake test              # Runs tests from the package defined in test.package property
    rake uninstall         # Uninstalls the application from a running emulator or device.

### Directory structure

Pindah follows the default Android directory structure when creating new projects:

    $ pindah create org.example.hello_world hello_world/ StartActivity

    Created 'hello_world/src/org/example/hello_world'.
    Created 'hello_world/bin'.
    Created 'hello_world/libs'.
    Created 'hello_world/res'.
    Created 'hello_world/res/drawable-hdpi'.
    Created 'hello_world/res/drawable-ldpi'.
    Created 'hello_world/res/drawable-mdpi'.
    Created 'hello_world/res/layout'.
    Created 'hello_world/res/values'.
    Created project in hello_world.

    $ cd hello_world && tree
    .
    |-- AndroidManifest.xml
    |-- libs
    |-- Rakefile
    |-- res
    |   |-- drawable-hdpi
    |   |   `-- ic_launcher.png
    |   |-- drawable-ldpi
    |   |   `-- ic_launcher.png
    |   |-- drawable-mdpi
    |   |   `-- ic_launcher.png
    |   |-- layout
    |   |   `-- main.xml
    |   `-- values
    |       `-- strings.xml
    `-- src
        `-- org
            `-- example
                `-- hello_world
                    `-- StartActivity.mirah

    12 directories, 8 files

Here are a few quick notes about the directory structure pictured above:

* The `AndroidManifest.xml` file describes the contents and metadata of your application.
* The `res/` directory contains application resources like icons, layout descriptions, and strings.
* The `src` directory contains the Java and Mirah source code for your project.

Other Android-specific directories may be created and used outside of this structure, as needed.  Please refer to the official Android documentation for more information about supported project file types and directory names.

### Compiling and running

To compile your new project in debug mode, simply run `rake debug` from your project directory:

    $ rake debug

    # [...]

    $ ls -l bin/hello_world-debug.apk

    -rw-r--r--   1 user           user        13222 Feb  7 23:16 bin/hello_world-debug.apk

This .apk file may be installed on a connected device or emulator with `rake installd`. It may even distributed for users to install themselves, though stable versions should be built with the `release` task.

The official documentation has [more details on building](http://developer.android.com/guide/developing/other-ide.html#Building). The main difference between Pindah and the standard Ant build is that the `Rakefile` replaces `build.xml` as well as all the properties files. The Android Developer page above also explains how to get an emulator running for testing your application.

See [Garrett](http://github.com/technomancy/Garrett) for an example of a basic project.

## Contributing

To contribute to Pindah, you need the [required environment](#requirements) setup. You can get tests with the following set of commands.

    $ git clone git@github.com:mirah/pindah.git
    $ bundle install
    $ rake test

## Community

Problems? Suggestions? Bring them up on the [Mirah mailing list](http://groups.google.com/group/mirah/) or the #mirah
IRC channel on freenode.

## See Also

If Mirah is just too low-level for your needs or you simply prefer something more dynamic, you can try [Ruboto](https://github.com/ruboto/ruboto/). Just be warned that it typically incurs a higher performance penalty and requires a larger app memory footprint due to JRuby's runtime.

## License

Released under the Apache 2.0 license.

Copyright (c) 2011 Phil Hagelberg, Nick Plante, J.D. Huntington



History.txt
=== 0.1.3 / 2014-01-15

* Resolve all outstanding bug reports in the issue tracker.
* Add support for the current Android API levels.
* Cleanup readme documentation formatting and update requirements,
  JRuby/Mirah combination compatibility, and usage information.
* Drop support for both JRuby 1.6.8 and Mirah 0.0.12.
* Bump minimum version requirements to JRuby 1.7.3 and Mirah 0.1.0.

=== 0.1.2 / 2011-06-06

* Silence Rake deprecation warnings.

=== 0.1.1 / 2011-04-02

* Declare package explicitly to work around Mirah bug.
* Make Activity creation optional.

=== 0.1.0 / 2011-02-08

* Birthday!


test/pindah_cli_test.rb
require 'test/unit'
require 'tmpdir'
require 'fileutils'
require 'rubygems'
require File.expand_path(File.join(File.dirname(__FILE__), '..', 'lib', 'pindah_cli'))

class PindahCLITest < Test::Unit::TestCase
  PWD = File.expand_path(File.dirname(__FILE__))

  def fixture(name)
    File.read(File.join(PWD, 'fixtures', name))
  end

  def setup
    @temp = Dir.mktmpdir("pindah-")
    @project_path = "#{@temp}/testapp"
    FileUtils.mkdir_p File.dirname(@temp)
    Dir.chdir File.dirname(@temp)
    PindahCLI.create('tld.pindah.testapp', @project_path, 'HelloWorld')
  end

  def teardown
    FileUtils.rm_rf @temp
  end

  def test_create_should_create_basic_project_structure
    PindahCLI.create('tld.pindah.testapp', '.')
    assert File.directory?(File.join(@project_path, 'src', 'tld', 'pindah', 'testapp'))

    directories = %w{ src/tld/pindah/testapp bin libs res
                      res/drawable-hdpi res/drawable-ldpi
                      res/drawable-mdpi res/layout res/values }

    directories.each do |d|
      expected = File.join(@project_path, d)
      assert File.directory?(expected), "Expected #{expected.inspect} to be a directory."
    end
  end

  def test_create_should_create_rakefile
    rake_path = File.join(@project_path, 'Rakefile')

    assert File.exists?(rake_path)
    assert_equal fixture("Rakefile"), File.read(rake_path)
  end

  def test_create_should_create_an_activity_if_desired
    actual = File.read(File.join(@project_path, 'src',
                                 'tld', 'pindah',
                                 'testapp', 'HelloWorld.mirah'))
    assert_equal fixture('HelloWorld.mirah'), actual
  end

  def test_create_should_create_manifest
    manifest_path = File.join(@project_path, 'AndroidManifest.xml')

    assert File.exists?(manifest_path)
    assert_equal fixture("AndroidManifest.xml").gsub(/\s+/, ' '), File.read(manifest_path).gsub(/\s+/, ' ')
  end

  def test_create_should_create_manifest_without_activity
    @temp = Dir.mktmpdir("pindah-")
    @project_path = "#{@temp}/testapp"
    FileUtils.mkdir_p File.dirname(@temp)
    Dir.chdir File.dirname(@temp)
    PindahCLI.create('tld.pindah.testapp', @project_path)

    manifest_path = File.join(@project_path, 'AndroidManifest.xml')

    assert File.exists?(manifest_path)
    assert_equal fixture("AndroidManifest.xml.no-activity").gsub(/\s+/, ' '), File.read(manifest_path).gsub(/\s+/, ' ')
  end

  def test_create_should_create_strings
    path = File.join(@project_path, 'res', 'values', 'strings.xml')
    assert File.exists?(path)
    assert_equal fixture("strings.xml").strip, File.read(path).strip
  end
  
  def test_create_should_create_layout
    path = File.join(@project_path, 'res', 'layout', 'main.xml')
    assert File.exists?(path)
    assert_equal fixture("main.xml").strip, File.read(path).strip
  end
end


test/pindah_build_test.rb
require 'test/unit'
require 'find'
require 'rubygems'
require 'fileutils'
require File.expand_path(File.join(File.dirname(__FILE__), '..',
                                   'lib', 'pindah'))

def Pindah.ant
  @ant
end

class PindahBuildTest < Test::Unit::TestCase
  def setup
    @garrett_path = File.expand_path(File.join(File.dirname(__FILE__), "garrett"))
    unless File.exists? @garrett_path
      system "git clone git://github.com/technomancy/Garrett.git #{@garrett_path}"
    end
    system "cd #{@garrett_path}; git pull"
  end

  def test_build
    cmd = "cd #{@garrett_path}; rake debug"
    assert system(cmd), "command failed: '#{cmd}'"

    ["Garrett-debug-unaligned.apk",
     "Garrett-debug.apk", "classes.dex"].each do |f|
      assert File.exists?("#{@garrett_path}/bin/#{f}"), "#{@garrett_path}/bin/#{f} didn't exist"
    end
    ["Garrett", "GarrettView", "R", "R$attr", "R$drawable",
     "R$layout", "R$string"].each do |c|
      assert File.exists? "#{@garrett_path}/bin/classes/garrett/p/serviss/#{c}.class"
    end
  end
end


lib/pindah_cli.rb
require 'fileutils'
require 'erb'

module PindahCLI
  DEFAULT_TARGET_VERSION = '2.3.3'

  def self.create(package, location=nil, activity_name=nil)
    segments = package.split('.')
    location ||= segments.last
    src_dir  = File.join(location, 'src', *segments)

    mkdir location, File.join('src', *segments)
    mkdir location, 'bin'
    mkdir location, 'libs'
    mkdir location, 'res'
    mkdir location, 'res/drawable-hdpi'
    mkdir location, 'res/drawable-ldpi'
    mkdir location, 'res/drawable-mdpi'
    mkdir location, 'res/layout'
    mkdir location, 'res/values'

    name = File.basename(location)
    app_name = name.split(/[-_]/).map{ |w| w.capitalize }.join(" ")

    FileUtils.mkdir_p src_dir

    create_templated("Rakefile", location, binding)
    create_templated("AndroidManifest.xml", location, binding)
    create_templated("main.xml", File.join(location, 'res', 'layout'), binding)
    create_templated("strings.xml", File.join(location, 'res', 'values'), binding)
    create_templated(".gitignore", location, binding)

    # Default icons of various sizes
    ["hdpi", "mdpi", "ldpi"].each do |s|
      FileUtils.cp(File.join(File.dirname(__FILE__), '..', 'templates', "res",
                             "drawable-#{s}", "ic_launcher.png"),
                   File.join(location, "res", "drawable-#{s}", "ic_launcher.png"))
    end
    
    log "Created project in #{location}."

    if activity_name
      activity_location = File.join(src_dir, "#{activity_name}.mirah")
      activity_template = File.read(File.join(File.dirname(__FILE__),
                                              '..', 'templates',
                                              'initial_activity.mirah'))

      File.open(activity_location, 'w') do |f|
        template_with_classname = activity_template.gsub(/INITIAL_ACTIVITY/, activity_name)
        f.puts ERB.new(template_with_classname).result(binding)
      end
      log "Created Activity '#{activity_name}' in '#{activity_location}'."
    end
  end

  private

  def self.log(msg)
    STDERR.puts msg
  end

  def self.create_templated(name, project_location, scope)
    location = File.join(project_location, name)
    template = File.read(File.join(File.dirname(__FILE__), 
                                   '..', 'templates', name))

    File.open(location, 'w') do |f|
      f.puts ERB.new(template).result(scope)
    end
  end

  def self.mkdir(base, directory)
    location = File.join(base, directory)
    FileUtils.mkdir_p location
    log "Created '#{location}'."
  end
end


lib/pindah.rb
require "rake"
require "fileutils"
require "tempfile"
require "pp"
require "erb"
require "pathname"

begin
  require 'ant'
rescue LoadError
  abort 'This Rakefile requires JRuby. Please use jruby -S rake.'
end

require "mirah"

module Pindah
  class << self
    include Rake::DSL
  end

  VERSION = '0.1.3'

  def self.infer_sdk_location(path)
    return ENV["ANDROID_HOME"] unless ENV["ANDROID_HOME"].nil?

    tools = path.split(File::PATH_SEPARATOR).detect { |p| File.exists?("#{p}/android") || File.exists?("#{p}/android.bat") }
    abort "\"android\" executable not found on $PATH" if tools.nil?
    real_location = Pathname.new("#{tools}/android").realpath.dirname
    File.expand_path("#{real_location}/..")
  end

  DEFAULTS = {
    :output => File.expand_path("bin"),
    :src => File.expand_path("src"),
    :classpath => Dir["jars/*jar", "libs/*jar"],
    :sdk => Pindah.infer_sdk_location(ENV["PATH"])
  }

  TARGETS = {
    "1.5"   => 3,
    "1.6"   => 4,
    "2.1"   => 7,
    "2.2"   => 8,
    "2.3"   => 9, "2.3.1" => 9, "2.3.2" => 9,
    "2.3.3" => 10, "2.3.4" => 10,
    "3.0"   => 11,
    "3.1"   => 12,
    "3.2"   => 13,
    "4.0"   => 14, "4.0.1" => 14, "4.0.2" => 14,
    "4.0.3" => 15, "4.0.4" => 15,
    "4.1"   => 16, "4.1.1" => 16, "4.1.2" => 16,
    "4.2"   => 17, "4.2.2" => 17,
    "4.3"   => 18,
    "4.4"   => 19
  }

  ANT_TASKS = ["javac", "compile", "clean", "debug", "release",
               "release_unsigned", "instrument", "test", "install", "installd",
               "installr", "installi", "installt", "uninstall"]
  SIGNED_TASKS = ["release"]
  ANT_TASK_MAP = (Hash.new { |h, k| h[k] = k }).merge({ "release_unsigned" => "release" })

  task :generate_manifest # TODO: generate from yaml?

  desc "Tail logs from a device or a device or emulator"
  task :logcat do
    system "adb logcat #{@spec[:log_spec]}"
  end

  desc "Print the project spec"
  task :spec do
    pp @spec
  end

  task :default => [:install]

  def self.spec=(spec)
    abort "Must provide :target_version in Pindah.spec!" if !spec[:target_version]
    abort "Must provide :name in Pindah.spec!" if !spec[:name]

    @spec = DEFAULTS.merge(spec)

    @spec[:root] = File.expand_path "."
    @spec[:target] ||= TARGETS[@spec[:target_version].to_s.sub(/android-/, '')]
    @spec[:classes] ||= "#{@spec[:output]}/classes/"
    @spec[:classpath] << @spec[:classes]

    if @spec.has_key?(:libraries)
      @spec[:libraries].each do |path|
        # TODO: do libraries always build to bin/classes?
        @spec[:classpath] << "#{File.expand_path path}/bin/classes/"
      end
    end

    @spec[:classpath] << "#{@spec[:sdk]}/platforms/android-#{@spec[:target]}/android.jar"
    @spec[:log_spec] ||= "ActivityManager:I #{@spec[:name]}:D " +
      "AndroidRuntime:E *:S"

    ant_setup
  end

  def self.ant_setup
    @ant = Ant.new

    # TODO: this is lame, but ant interpolation doesn't work for project name
    build_template = ERB.new(File.read(File.join(File.dirname(__FILE__), '..',
                                                 'templates', 'build.xml')))
    @build = Tempfile.new(["pindah-build", ".xml"])
    @build.write(build_template.result(binding))
    @build.close

    if @spec.has_key?(:libraries)
      # Push the libs in through project.properties
      props_template = ERB.new(File.read(File.join(File.dirname(__FILE__), '..',
                                                   'templates', 'project.properties')))
      @props = File.new("project.properties", 'w')
      @props.write(props_template.result(binding))
      @props.close
    end

    user_properties = {
      "target" => "android-#{@spec[:target]}",
      "target-version" => "android-#{@spec[:target_version]}",
      "src" => @spec[:src],
      "sdk.dir" => @spec[:sdk],
      "classes" => @spec[:classes],
      "classpath" => @spec[:classpath].join(Java::JavaLang::System::
                                            getProperty("path.separator"))
    }

    user_properties.each do |key, value|
      @ant.project.set_user_property(key, value)
    end

    Ant::ProjectHelper.configure_project(@ant.project, java.io.File.new(@build.path))

    # Turn ant tasks into rake tasks
    add_tasks(ANT_TASKS)
    add_tasks(@spec[:extra_tasks])
  end

  protected

  def self.add_tasks(tasks=nil)
    return if tasks.nil?
    tasks.each do |name|
      ant_name = ANT_TASK_MAP[name]

      target_task = @ant.project.targets[ant_name]

      if target_task && target_task.description
        desc target_task.description.gsub(/\s\s+/, "\n")
      else
        desc "Run #{name}"
      end

      task(name) do
        add_signature_properties if SIGNED_TASKS.include?(name)
        @ant.project.execute_target(ant_name)
      end
    end
  end

  def self.add_signature_properties
    # Add key signing config
    if @spec[:key_store] && @spec[:key_alias]

      # NB: due to a JRuby/Ant bug, Ant can't read these passwords from the
      # command line.
      #
      # So, we'll work around this for now. Icky.
      #
      # See: http://jira.codehaus.org/browse/JRUBY-4827
      puts "Please enter keystore password (store:#{@spec[:key_store]}):"
      store_pw = STDIN.gets.chomp

      puts "Please enter password for alias '#{@spec[:key_alias]}':"
      alias_pw = STDIN.gets.chomp

      signature_properties = {
        "key.store" => @spec[:key_store],
        "key.alias" => @spec[:key_alias],
        "key.store.password" => store_pw,
        "key.alias.password" => alias_pw
      }

      signature_properties.each do |key, value|
        @ant.project.set_user_property(key, value)
      end

      # NB: we need to do this again to actually set the new properties.
      Ant::ProjectHelper.configure_project(@ant.project, java.io.File.new(@build.path))
    end
  end
end

